sway-core/src/semantic_analysis/ast_node/expression/typed_expression/function_application.rs

```diff
@@ -107,59 +120,71 @@ pub(crate) fn instantiate_function_application_simple(
-    function_decl: TypedFunctionDeclaration,
-    selector: Option<ContractCallMetadata>,
-    is_constant: IsConstant,
-    self_state_idx: Option<StateIndex>,
-    span: Span,
-) -> CompileResult<TypedExpression> {
+fn check_function_arguments_arity(
+    arguments_len: usize,
+    function_decl: &TypedFunctionDeclaration,
+    call_path: &CallPath,
+) -> CompileResult<()> {
     let warnings = vec![];
     let mut errors = vec![];
-    match arguments.len().cmp(&function_decl.parameters.len()) {
-        std::cmp::Ordering::Equal => {
-            let exp = TypedExpression {
-                expression: TypedExpressionVariant::FunctionApplication {
-                    call_path,
-                    contract_call_params,
-                    arguments,
-                    function_body: function_decl.body.clone(),
-                    function_body_name_span: function_decl.name.span(),
-                    function_body_purity: function_decl.purity,
-                    self_state_idx,
-                    selector,
-                },
-                return_type: function_decl.return_type,
-                is_constant,
-                span,
-            };
-            ok(exp, warnings, errors)
-        }
+    match arguments_len.cmp(&function_decl.parameters.len()) {
+        std::cmp::Ordering::Equal => ok((), warnings, errors),
         std::cmp::Ordering::Less => {
             errors.push(CompileError::TooFewArgumentsForFunction {
-                span,
-                method_name: function_decl.name,
+                span: call_path.span(),
+                method_name: function_decl.name.clone(),
                 expected: function_decl.parameters.len(),
-                received: arguments.len(),
+                received: arguments_len,
             });
             err(warnings, errors)
         }
         std::cmp::Ordering::Greater => {
             errors.push(CompileError::TooManyArgumentsForFunction {
-                span,
-                method_name: function_decl.name,
+                span: call_path.span(),
+                method_name: function_decl.name.clone(),
                 expected: function_decl.parameters.len(),
-                received: arguments.len(),
+                received: arguments_len,
             });
             err(warnings, errors)
         }
     }
 }
+
+#[allow(clippy::too_many_arguments)]
+#[allow(clippy::comparison_chain)]
+fn instantiate_function_application_inner(
+    call_path: CallPath,
+    contract_call_params: HashMap<String, TypedExpression, RandomState>,
+    arguments: Vec<(Ident, TypedExpression)>,
+    function_decl: TypedFunctionDeclaration,
+    selector: Option<ContractCallMetadata>,
+    is_constant: IsConstant,
+    self_state_idx: Option<StateIndex>,
+    span: Span,
+) -> TypedExpression {
+    assert_eq!(
+        arguments.len(),
+        function_decl.parameters.len(),
+        "different number of function parameters and arguments"
+    );
```

Comment:
We tend not to have assertions in the compiler as permanent code (sometimes there'll be a `todo!()` or `unimplemented!()` temporarily) but instead prefer to return a `CompileError::Internal` to encourage users to file an issue if something like this occurs.